quickinput-sample/src/promptCommandWithHistory.ts

```diff
@@ -0,0 +1,101 @@
+import { window, Disposable, Memento } from 'vscode';
+import { QuickPickItem } from 'vscode';
+
+const HISTORY_KEY = 'HISTORY_KEY';
+
+
+/**
+ * A command prompt with history
+ *
+ */
+export const getPromptCommand = (memo: Memento) => {
+	const pickCommand = getPickCommand(memo);
+	// §TODO: cache promptCommand!
+	return async function promptCommand() {
+		const command = await pickCommand();
+		if (command) {
+			window.showInformationMessage(`You picked the following command: '${command}'`)
+		}
+	}
+}
+
+class CommandItem implements QuickPickItem {
+	public static HISTORY: string = 'history';
+	public static INPUT: string = 'input';
+
+	public label: string;
+	public description?: string;
+	public type: string;
+
+	constructor(type: string, label: string, description?: string) {
+		this.type = type;
+		this.label = label;
+		this.description = description;
+	}
+
+	static input(content: string): CommandItem {
+		return new CommandItem(CommandItem.INPUT, content, '(current input)');
+	}
+	static history(content: string, description?: string): CommandItem {
+		return new CommandItem(CommandItem.HISTORY, content, description);
+	}
+}
+
+const getPickCommand = (memo: Memento) => async function pickCommand() {
+	const disposables: Disposable[] = [];
+	try {
+		return await new Promise<string | undefined>((resolve, reject) => {
+			const input = window.createQuickPick<CommandItem>();
+			input.placeholder = 'Type a command';
+			input.items = [];
+
+			const updateQuickPick = (value?: string): void => {
+				if (value === input.value) return;
+				if (!value) {
+					if (input.items[0] && input.items[0].type === 'input')
+						input.items = input.items.slice(1);
+					return;
+				}
+				input.items = [CommandItem.input(value)].concat(
```

Comment:
Couldn't you use `input.items = [CommandItem.input(value)].concat(historyItems)` when you move `historyItems` before `updateQuickPick`? And `input.items = historyItems` above. Looks good otherwise.

Reply:
indeed, and it got far better!